ENH: Add `scipy.special._ufuncs._iv_ratio`

fancidev opened the pull request:
#### Reference issue
Implement part of gh-20253.

#### What does this implement/fix?
Adds a private function `_iv_ratio` to compute `iv(v,x)/iv(v-1,x)` for `v >= 1` and `x >= 0`. One of `v` or `x` may be `inf` but not both.

For inputs in the domain, return a real number between `0` and `1`, inclusive. For inputs out of the domain, set error and return `nan`.

#### Algorithm

Let $R_\nu(x) := I_\nu(x) / I_{\nu-1}(x)$ for $\nu \ge 1$ and $x > 0$. The below results are also valid for $x = 0$.

**Continued fraction**.  To compute $R_\nu(x)$, we use the _Perron continued fraction_ (Gautschi and Slavik, 1978)

$$
R_\nu(x) = \frac{x}{x+f(\nu,x)}
$$

where

$$
f(\nu,x) = 2\nu + \frac{-(2\nu+1)x}{2(\nu+x)+1+{}} \frac{-(2\nu+3)x}{2(\nu+x)+2+{}} \frac{-(2\nu+5)x}{2(\nu+x)+3+{}} \cdots
$$

<!--
Since $0 < R_\nu(x) < 1$, we have $f(\nu,x) > 0$.
-->

The following bounds of $f(\nu,x)=x/R_\nu(x)-x$ prove useful. Amos (1974, Eq. 16) showed that

$$
\frac{x}{\nu-0.5+\sqrt{x^2+(\nu+0.5)^2}}\le R_\nu(x)\le\frac{x}{\nu-0.5+\sqrt{x^2+(\nu-0.5)^2}}
$$

It follows that

$$
\nu-0.5-x+\sqrt{x^2+(\nu-0.5)^2} \le 
f(\nu,x) \le \nu-0.5-x+\sqrt{x^2+(\nu+0.5)^2}
$$

The bounds may be relaxed to

$$
\nu-0.5 \le f(\nu,x) \le 2\nu
$$

**Evaluation of the continued fraction**.  Given a continued fraction $f$, let

$$
f_n = b_0 + \frac{a_1}{b_1+{}} \frac{a_2}{b_2+{}} \cdots \frac{a_n}{b_n+0}
$$

denote its $n$-th convergent.  It can be shown (Gautschi, 1967) that for $k \ge 2$,

$$
\begin{align*}
\frac{f_k-f_{k-1}}{f_{k-1}-f_{k-2}} = \rho_k &:= -\left(\frac{b_k b_{k-1}}{a_k (1+ \rho_{k-1})}+1\right)^{-1}
\end{align*}
$$

with $\rho_1 \equiv 0$. Then, for $n \ge 1$, rearrange $f_n$ as

$$
\begin{align*}
f_n &= f_0 + \sum_{k=1}^n (f_k-f_{k-1}) \\
&= f_0 + \sum_{k=1}^n \frac{a_1}{b_1} \left( \rho_2 \rho_3 \cdots \rho_k \right)
\end{align*}
$$

Since $\lim_{n\rightarrow \infty} f_n =f$, $f$ is then evaluated by $f_n$, $n \ge 1$ such that $| f_n-f_{n-1} | \le \epsilon |f_n|$ for some prescribed relative tolerance $\epsilon > 0$ (the “termination condition”).

For the Perron continued fraction in question, we have

$$
\begin{align*}
b_0 &= 2\nu \\
a_k &= -2\nu x-(2k-1)x \\
b_k &= 2\nu+2x+k
\end{align*}
$$

for $k \ge 1$.

**Monotonicity of the convergents**.  Direct calculation shows that

$$
\rho_k = \frac{1+\rho_{k-1}}{2h(k;\nu,x)+(1-\rho_{k-1})}
$$

where

$$
h(k;\nu,x) := \frac{1}{4x}\left[ \left(\nu+k-0.5\right) +\frac{ (\nu+2x-0.5)(\nu+2x+0.5)}{\nu+k-0.5}+2\nu \right]
$$

It is easy to see that

$$
\begin{align*}
h(k;\nu,x) &\ge \frac{1}{4x}\left[ 2\sqrt{ (\nu+2x-0.5)(\nu+2x+0.5) } +2\nu \right] \\
&\ge \frac{1}{4x}\left[ 2(\nu+2x-0.5) +2\nu \right] \\
&=\frac{\nu+x-0.25}{x} \\
&:=\bar{h}(\nu,x)
\end{align*}
$$

where the lower bound in the first inequality is attained if $k=k^\ast(\nu,x)$ where $\nu+k^\ast-0.5=\sqrt{(\nu+2x-0.5)(\nu+2x+0.5)}$; this implies $2x < k^\ast < 2x+1$. 

Starting from $\rho_1 \equiv 0$ and using the recurrence relation $\rho_k \le (1+\rho_{k-1}) / [2\bar{h}(\nu,x)+(1-\rho_{k-1})]$, it is straightforward to verify by induction that

$$
0 \le \rho_k \le \bar\rho(\nu,x) := \frac{x}{2\nu +x-0.5}  < 1
$$

for all $k \ge 1$. Since $b_0 > 0$, $a_1/b_1 < 0$, and $0 < \rho_k < 1$ for $k \ge 2$, we conclude that $f_n$ is monotonically decreasing in $n$, and that $\nu-0.5 \le f \le f_n \le  b_0=2\nu$ for all $n \ge 1$.

**Equivalence transform**.  When evaluating $a_1/b_1$ and $\rho_k$, intermediate calculations may overflow if $\nu$ and/or $x$ is large. Such overflow can be avoided by applying an equivalence transform of the continued fraction as follows. Given any constant $c \ne 0$, rewrite $R_\nu(x)$ as

$$
R_\nu(x) = \frac{xc}{xc+f(\nu,x)c}
$$

where

$$
f(\nu,x)c = 2\nu c  + \frac{-(2\nu c+c)(xc)}{2(\nu c+x c)+c+{}} \frac{-(2\nu c+3c)(xc)}{2(\nu c+xc)+2c+{}} \frac{-(2\nu c+5c)(xc)}{2(\nu c+xc)+3c+{}} \cdots
$$

Specifically, we choose $c=2^{2-p}$ where $p$ is the integer that satisfies $2^{p-1} \le \max(\nu,x) < 2^p$. For the range of double precision numbers greater than or equal to one, $p$ ranges from $1$ to $1024$, and $c$ ranges from $2$ to $2^{-1022}$. It is easy to verify that the transformed continued fraction can be evaluated without triggering floating point overflow in any intermediary steps. In addition, the values of $\rho_k$, $k \ge 1$ are unchanged by the transform, while $b_0$ and $a_1/b_1$ are scaled by $c$.

**Truncation error**.  The termination condition is necessary but not sufficient for convergence. In fact, Gautschi (1977) showed that it is prone to early termination for complex $x$. It is therefore useful to establish a bound for the truncation error.

For $n \ge 1$, we have

$$
\begin{align*}
|f-f_n| &= \left|\sum_{k=n+1}^{\infty} \frac{a_1}{b_1} ( \rho_2 \cdots \rho_k ) \right|\\
&= \left|\frac{a_1}{b_1} ( \rho_2 \cdots \rho_n )\right| \sum_{k=n+1}^\infty ( \rho_{n+1} \cdots \rho_k ) \\
&= | f_n-f_{n-1}| \sum_{k=n+1}^\infty ( \rho_{n+1} \cdots \rho_k )  \\
&\le |f_n-f_{n-1}| \sum_{k=n+1}^{\infty} {\bar\rho}^{k-n} \\
&= |f_n-f_{n-1}|\frac{\bar\rho}{1-\bar\rho}
\end{align*}
$$

It follows that when $R_\nu(x)$ is evaluated by $x/(x+f_n)$ and the termination condition $|f_n-f_{n-1}| \le \epsilon |f_n|$ is satisfied, the relative truncation error is bounded by

<!--

$$
\begin{align*}
\frac{|f_n-f|}{|f|} 
&\le \frac{|f_n-f_{n-1}|}{|f|}\frac{\bar\rho}{1-\bar\rho} \\
&\le \frac{|f_n|}{|f|} \frac{\epsilon\bar\rho}{1-\bar\rho} \\
&\le \frac{2\nu}{\nu-0.5} \frac{\epsilon\bar\rho}{1-\bar\rho} 
%&= \frac{\epsilon x}{\nu-0.25}
\end{align*}
$$

-->

$$
\begin{align*}
\left| \frac{x/(x+f_n)}{x/(x+f)} - 1\right|
&= \left| \frac{f-f_n}{x+f_n} \right| \\
&\le \frac{|f_n-f_{n-1}|}{x+f_n} \frac{ \bar\rho}{1-\bar\rho} \\
&\le \frac{f_n}{x+f_n} \frac{\epsilon \bar\rho}{1-\bar\rho} \\
&\le \frac{2\nu}{x+2\nu} \frac{\epsilon x}{2\nu-0.5} \\
&\le \frac{4}{3} \epsilon \\
\end{align*}
$$

<!--

On the other hand, when $1-R_\nu(x)$ is evaluated by $f_n/(x+f_n)$, the relative truncation error bounded by

$$
\left| \frac{f_n/(x+f_n)}{f/(x+f)} - 1\right|
= \frac{x}{x+f_n} \left| \frac{f_n-f}{f} \right|
\le  \frac{\epsilon\bar\rho}{1-\bar\rho} 
= \frac{\epsilon x}{\nu-0.25}
$$

This bound is unfortunately too coarse if $x \gg \nu$.

-->

Note that the relative truncation error is not affected by an equivalence transform.

**Rate of convergence**. We now derive a bound on the number of terms required for the termination condition. Substituting $f_n-f_{n-1}=(a_1/b_1)(\rho_2 \cdots \rho_n)$ into the termination condition yields

$$
\left| \frac{f_n-f_{n-1}}{f_n} \right| = \frac{1}{f_n} \left| \frac{a_1}{b_1} \right| (\rho_2 \cdots \rho_n)
$$

Since $f_n \ge f \ge \nu-0.5$, the first term is bounded by

$$
\begin{align*}
\frac{1}{f_n} \left| \frac{a_1}{b_1} \right|
&\le \frac{(2\nu+1)x}{(2\nu+2x+1)(\nu-0.5)} %\\
%&\le \frac{2\nu+1}{2(\nu-0.5)} \\
\le 3
\end{align*}
$$

The second term is bounded by ${\bar\rho}^{n-1}$, but that bound is too coarse for $x \gg \nu$. To obtain a sharper bound, rearrange $\rho_k$, $k \ge 2$ as

$$
\rho_k = \frac{1+\rho_{k-1}}{2 g(\nu;x,k)+(1-\rho_{k-1})}
$$

where

$$
g(\nu;x,k) := 1 + \frac{1}{x} \left[\nu+\frac{ (x-\frac{k}{2})(x-\frac{k}{2}+\frac{1}{2})}{\nu+k-\frac{1}{2}} \right]
$$

Since

$$
g'(\nu) = \frac{1}{x} \left[1-\frac{ (x-\frac{k}{2})(x-\frac{k}{2}+\frac{1}{2})}{\left(\nu+k-\frac{1}{2}\right)^2} \right]
$$

we can see that $g(\nu)$ has exactly one minimum over $\nu \ge 1$, which is attained at $\nu=1$ if $g'(1) \ge 0$, or at some $\nu^\ast > 1$ such that $g'(\nu^\ast)=0$ if $g'(1) < 0$.

A sufficient condition for $g'(1) \ge 0$ is

$$
\begin{align*}
&(x-\frac{k}{2})(x-\frac{k}{2}+\frac{1}{2}) \le 0 \\
\textrm{or   } &x-\frac{k}{2} \ge 0 \textrm{  and  } k+\frac{1}{2} \ge x-\frac{k}{2}+\frac{1}{2} \\
\textrm{or   } &x-\frac{k}{2}+\frac{1}{2} \le 0 \textrm{  and  } k+\frac{1}{2} \ge \frac{k}{2}-x-\frac{1}{2}
\end{align*}
$$

which can be combined into $x \le \frac{3}{2} k$. In this case, $g(\nu) \ge g(1)$ where

$$
g(1) = \frac{1}{2k+1}\left[ 2x+\frac{(k+1)(k+2)}{2x}+2\right]
$$

The minimum of $g(1;x,k)$ is attained at  $x^\ast(k)=\frac{1}{2}\sqrt{(k+1)(k+2)}$.

Similarly, it can be shown that a sufficient condition for $g'(1) \le 0$ is $x \ge \frac{1}{2} + \frac{3}{2} k$. In this case, $g(\nu;x,k) \ge g(\nu^\ast;x,k)$ where

$$
\begin{align*}
g(\nu^\ast;x,k)
&=1+\frac{1}{x}\left[2 \sqrt{\left(x-\frac{k}{2}\right)\left(x-\frac{k}{2}+\frac{1}{2}\right)} -k+\frac{1}{2}\right]\\
&\ge 1+\frac{1}{x}\left[ 2 \left(x-\frac{k}{2}\right)-k+\frac{1}{2}\right]\\
&=3-\frac{4k-1}{2x} \\
&:= \bar{g}(x,k)
\end{align*}
$$

Note that the above in fact holds for all $x \ge \frac{k}{2}$, though the bound is not sharp if $x \le \frac{3}{2}k$.

We proceed to bound $\rho_k$ for “large” and “small” $x$ separately. Let $N$ be a suitably chosen positive integer. First  consider $x \ge N$ and $2 \le k \le N$. Under this parameter range, $g(\nu;x,k) \ge g(\nu^\ast;x,k) \ge \bar{g}(x,k) \ge \bar{g}(N,k)$. It follows that

```math
\rho_k \le {\bar\rho}_k := \frac{1+{\bar\rho}_{k-1}}{2\bar{g}(N,k)+(1 -{\bar\rho}_{k-1})}
```

with ${\bar\rho}_1 \equiv 0$, and

```math
\rho_2\cdots\rho_n \le {\bar\rho}_2 \cdots {\bar\rho}_n
```

provided $n \le N$.

Next, consider $x \le N$ and $k \ge 2N$. Under this parameter range, $g(\nu;x,k) \ge g(1;x,k) \ge g(1;N,k)$, as $x^\ast(k) > N$ falls out of the domain. It follows that

```math
\rho_k \le {\tilde\rho}_k := \frac{1+{\tilde\rho}_{k-1}}{2{g}(1;N,k)+(1 -{\tilde\rho}_{k-1})}
```

with ${\tilde\rho}_{2N-1} := \bar{\rho}(1,N)$, and

```math
\rho_2\cdots\rho_n \le \bar{\rho}(1,N)^{2N-2} {\tilde\rho}_{2N} \cdots {\tilde\rho}_n
```

provided $n \ge 2N$.

<details>

<summary>
The following script computes a bound on the number of terms required to achieve a given tolerance.
</summary>

```python
import numpy as np

eps = np.finfo(float).eps / 2
N = 32

p = 0
s = 1
for k in range(2, N+1):
    g = 3-(4*k-1)/(2*N)
    p = (1+p)/(2*g+(1-p))
    s *= p
    if s < eps/3:
        print(f'For x <= {N}, p[2]...p[{k}] = {s}')
        break

p = N/(N+2-0.5)
s = p**(2*N-2)
for k in range(2*N, 10000):
    g = (2*N+(k+1)*(k+2)/(2*N)+2) / (2*k+1)
    p = (1+p)/(2*g+(1-p))
    s *= p
    if s < eps/3:
        print(f'For x >= {N}, p[2]...p[{k}] = {s}')
        break
```

</details>

For double precision, we set $\epsilon = 2^{-53} \approx 1.11\times 10^{-16}$, and choose $N=32$. For $x\ge 32$, we find $\rho_2 \cdots \rho_{31} \le 3.14 \times 10^{-17}$. For $x\le 32$, we find $\rho_2\cdots \rho_{137} \le 2.00\times 10^{-17}$. Hence the number of terms is bounded by $137$.

#### Reference

Amos, D. E. (1974). "Computation of Modified Bessel Functions and Their Ratios." _Mathematics of Computation_, 28(125):239-251. [Link](https://www.jstor.org/stable/i334548)

Gautschi, W. (1967). “Computational Aspects of Three-Term Recurrence Relations.” _SIAM Review_, 9(1):24-82. [Link](https://www.jstor.org/stable/i309438)

Gautschi, W. (1977).  "Anomalous Convergence of a Continued Fraction for Ratios of Kummer Functions." _Mathematics of Computation_, 31(140):994-999. [Link](https://www.jstor.org/stable/i308530)

Gautschi, W. and Slavik, J. (1978). "On the computation of modified Bessel function ratios." _Mathematics of Computation_, 32(143):865-875. [PDF](https://www.ams.org/journals/mcom/1978-32-143/S0025-5718-1978-0470267-9/S0025-5718-1978-0470267-9.pdf)

#### Additional information

It is known that $I_{1.5}(x)/I_{0.5}(x) \equiv 1/{\tanh x} - 1/x$. However, the latter expression has to be evaluated using series expansion for moderately small $x$ (such as $x < 0.1$) to avoid catastrophic cancellation error.

steppi committed:
Add continued fraction evaluator

steppi committed:
Add stub for iv_ratio

fancidev committed:
Basic implementation of iv_ratio.

fancidev committed:
Polish the continued fraction evaluator.

izaid commented:
@fancidev Thanks for submitting this! I'm happy to throw out a couple of points now, hopefully helpful. I'm all for increasing coverage of special functions, and great to see additions at the C++ level. Some thoughts:

- We are moving away from `functions.json`. I know the documentation says that is the way to add special functions, but that's only because we haven't gotten rid of it yet (and didn't expect contributions of new special functions). The way to add special functions now is to add an entry in `_special_ufuncs.cpp` with corresponding documentation in `_special_ufuncs_docs.cpp`. See [here](https://github.com/scipy/scipy/blob/main/scipy/special/_special_ufuncs.cpp#L327) for `jv` as an example, it's pretty much copy and paste. Then add the name to `special_ufuncs` in `_generate_pyx.py`, see [here](https://github.com/scipy/scipy/blob/main/scipy/special/_generate_pyx.py#L78).

- Regarding 'maybe_complex_numeric_limits', I don't think we want that. What I would suggest is adding something like:
```
template <typename T>
struct real_type {
    using type = T;
};

template <typename T>
struct real_type<std::complex<T>> {
    using type = T;
};

template <typename T>
using real_type_t = typename real_type::type;
```
Then you can do `std::numeric_limits<real_type_t<T>>` and it will work regardless of whether `T` is complex or not.

That's it for the moment, but more thoughts as we go.

steppi commented:
Thanks @fancidev and thanks @izaid for taking a look. I'll be able to take a look on Monday.

> * Regarding 'maybe_complex_numeric_limits', I don't think we want that. What I would suggest is adding something like:

Sorry. That abomination is my fault. @fancidev's work started with the continued fraction evaluator that I wrote which included that nonsense.

> * We are moving away from `functions.json`.

I personally think it's fine for new functions to use the old infrastructure while things are still in transition, and before we document the new process. Things can always be updated later. Also, @izaid, I think one of us should help out with manual update to Cython special needed after https://github.com/scipy/scipy/pull/20417. I was hoping to document all of these things next week.

steppi commented on scipy/special/special/tools.h:
```diff
@@ -4,6 +4,8 @@
 
 #include "config.h"
 #include "error.h"
+#include <utility>
+#include <cstddef>
```
All includes from the standard library need to go in [config.h](https://github.com/scipy/scipy/blob/main/scipy/special/special/config.h) and we use conditional compilation to swap between CUDA versions or standard library versions based on platform. For anything you use, you need to make sure that it is available in `libcudacxx` or `thrust`.

fancidev replied:
I use the headers for `std::pair` and `std::size_t`. Let me check those in `config.h`.

steppi replied:
@izaid and I have our own GPU workstations to use. Don't worry about this aspect for now. It should be straightforward to adjust things after you get everything working.

steppi commented on scipy/special/special/tools.h:
```diff
@@ -73,5 +75,273 @@ namespace detail {
+     * The numerical accuracy of this method depends on the characteristics of
+     * the continued fraction being evaluated.
+     */
+    template <typename T, typename InputIt>
+    SPECFUN_HOST_DEVICE std::pair<T, std::size_t> continued_fraction_eval_series(
+        T init_val, InputIt cf, T tol, std::size_t max_terms) {
```
Can we factor this so that we define a new iterator which takes an iterator for a continued fraction and tranforms it into an iterator for the terms of thid series, which can then be fed into a series evaluator, perhaps a regular one, a Kahan summation evaluator. We also want to add convergence acceleration transforms that series iterators could be passed into.

fancidev replied:
Sounds good. I can make something like `ContinuedFractionDifferenceSeries` to yield the adjacent difference between partial convergents. That will make the logic clearer. As for "without worrying about details", I'm afraid the caller still need to do some serious analysis. For example, a series evaluator can be as good as it is, but if the caller supplies a slow converging series, they can't expect it to run quickly!

steppi replied:
Cool. Understanding the math is needed, but I want people who know the math to be able to write code at a higher , almost declarative level, without needing to worry too much about programming details. `scipy.special` was unmaintained for long time because it requires knowledge both of the math and programming in compiled languages, and the hope is to simplify things for people who are stronger in the former.

fancidev commented:
Thanks @izaid and @steppi for the comments. Regarding code organisation: as `scipy.special` is going through some refactoring right now, I'll complete the other (functional) points first and then adapt to the best practice of `scipy.special` code structure.

fancidev commented:
Also a question @steppi : Since the series evaluator, etc are templated, do you expect them to support custom data types such as arbitrary-precision floats and `__float128`, or do you restrict the support to `float`, `double` and their complex version? I'm asking this because if you intend to support custom floats, then the type of `tol` would better be templated as well.

steppi commented:
> Also a question @steppi : Since the series evaluator, etc are templated, do you expect them to support custom data types such as arbitrary-precision floats and `__float128`, or do you restrict the support to `float`, `double` and their complex version? I'm asking this because if you intend to support custom floats, then the type of `tol` would better be templated as well.

Right now we're just supporting `float` and `double` and their complex versions. Longer term, we've talked about supporting extended precision arithmetic. I think `tol` should be templated though and should be the real type for whatever `T` is.

izaid commented:
> > Also a question @steppi : Since the series evaluator, etc are templated, do you expect them to support custom data types such as arbitrary-precision floats and `__float128`, or do you restrict the support to `float`, `double` and their complex version? I'm asking this because if you intend to support custom floats, then the type of `tol` would better be templated as well.
> 
> Right now we're just supporting `float` and `double` and their complex versions. Longer term, we've talked about supporting extended precision arithmetic. I think `tol` should be templated though and should be the real type for whatever `T` is.

Depends what is needed, but using `std::numeric_limits<T>::epsilon()` (https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon) for the tolerance may be enough.

fancidev commented:
Cool. I’ll make the type of tolerance `decltype(abs(T(0)))` and set its default value to (maybe half) `epsilon`.

Related question: do you prefer `< tol` or `<= tol` for termination? If you don’t care then I’ll go with `<=` because error bounds are typically written as `<=`, and this allows one to set `tol = 0` to run to the limit of the machine!

steppi commented:
I think this is how you'd want to type the tolerance. @izaid?

```C++
template <typename T, typename InputIt>
    SPECFUN_HOST_DEVICE std::pair<T, std::size_t> continued_fraction_eval_series(
        T init_val, InputIt cf, real_type_t<T> tol, std::size_t max_terms) {
```

izaid commented:
> I think this is how you'd want to type the tolerance. @izaid?
> 
> ```c++
> template <typename T, typename InputIt>
>     SPECFUN_HOST_DEVICE std::pair<T, std::size_t> continued_fraction_eval_series(
>         T init_val, InputIt cf, real_type_t<T> tol, std::size_t max_terms) {
> ```

Is `T` possibly complex? Then yes, that's how I'd do it.

steppi commented:
> Is `T` possibly complex? Then yes, that's how I'd do it.

Yes, possibly complex. We're talking generic series and continued fraction evaluators meant to work for both real and complex types.

> Related question: do you prefer `< tol` or `<= tol` for termination? If you don’t care then I’ll go with `<=` because error bounds are typically written as `<=`, and if a mathematically derived bound happens to round to zero in machine float, setting `tol = 0` will still work.

I prefer `< tol`. It usually wouldn't even matter, but if the relative error ends up as exactly machine epsilon, I suspect you'd more likely than not be better off trying one more iteration. I don't think there's ever a good reason to set `tol = 0` for relative error, and even if using `< tol`, setting it to something sufficiently small will effectively be the same as setting it to `0` with `<= tol`.

fancidev commented:
I’ll type tol using real_type.

For `<= tol` vs `< tol`, a practical context is the following. To compute `iv_ratio(v,x)`, to achieve a relative truncation error no larger than $\epsilon$ (say `1e-16`), it is sufficient to terminate iteration as soon as

$$
\frac{f_n-f_{n-1}}{f_n} \le \epsilon \cdot \frac{\nu+0.75}{x}
$$

where $f_n$ is the $n$-th convergent. It is then straightforward to set `tol` to the right-hand side of the above inequality. However, this will cause an infinite loop (or declared non-convergence) if the right-hand side is numerically zero (e.g. if `v=0` and `x=1e300`) and `< tol` is used as the termination criteria.

fancidev committed:
Refactor continued fraction convergent difference into a generator.

steppi commented:
@fancidev, just set  the tolerance to the max of your right hand side and $\epsilon$ or possibly $\epsilon/2$. I think this communicates intent more clearly, rather than relying on what should be an arbitrary property of how the stopping criterion is defined.

fancidev committed:
Refactor into Kahan series evaluator.

fancidev commented:
> @fancidev, just set the tolerance to the max of your right hand side and ϵ or possibly ϵ/2. I think this communicates intent more clearly, rather than relying on what should be an arbitrary property of how the stopping criterion is defined.

Ok!

Meanwhile I have refactored `continued_fraction_eval_series` into a Kahan series evaluator and an iterator that generates the difference between successive convergents. Thanks for taking a look!

On a completely separate topic, `hyp2f1.h` makes some compiler warnings about unused field, not sure if a problem:

```
../scipy/special/special/hyp2f1.h:428:16: warning: private field 'a_' is not used [-Wunused-private-field]
        double a_, b_, c_, m_;
               ^
```

steppi commented:
> On a completely separate topic, `hyp2f1.h` makes some compiler warnings about unused field, not sure if a problem:

That's just an oversight on my part, and should be fixed in a separate PR.

```C++
    class Hyp2f1Transform2LimitFinitePartGenerator {
        /* Initial finite sum in limit as a - b approaches a non-negative integer m. The limiting series
         * for the 1 - z transform also has an initial finite sum, but it is a standard hypergeometric
         * series. */
      public:
        SPECFUN_HOST_DEVICE Hyp2f1Transform2LimitFinitePartGenerator(double a, double b, double c, double m,
                                                                     std::complex<double> z)
            : a_(a), b_(b), c_(c), m_(m), z_(z), term_(cephes::Gamma(m) / cephes::Gamma(c - b)), k_(0) {}

        SPECFUN_HOST_DEVICE std::complex<double> operator()() {
            std::complex<double> output = term_;
            term_ = term_ * (b_ + k_) * (c_ - b_ - k_ - 1) / ((k_ + 1) * (m_ - k_ - 1)) / z_;
            ++k_;
            return output;
        }

      private:
        double a_, b_, c_, m_;
        std::complex<double> z_, term_;
        std::uint64_t k_;
    };
```

`a` isn't actually needed because I wrote things in terms of `m = a - b` and `b`.

steppi commented on scipy/special/_iv_ratio.h:
```diff
@@ -0,0 +1,118 @@
+    double tol = std::numeric_limits<double>::epsilon() * (v+0.75);
+    tol = tol / std::max(tol, x); // prevent overflow
+#if 1
+    auto [result, terms] = special::detail::continued_fraction_eval_series(
+        0.0, cf, tol, 1000);
```
I think it's more clear to just generate the terms and sum them here, rather than writing a separate `continued_fraction_eval_series` which just has those two lines. The idea is to have simple composable tools .

fancidev replied:
Done!

fancidev committed:
Clean up and add continued_fraction_series() function.

fancidev committed:
Refactor to use generator protocol.

fancidev committed:
Cosmetic changes to make code style more consistent.

fancidev committed:
Handle invalid arguments to _iv_ratio.

fancidev commented:
I made some further polishing. In particular, I added a templated `next` function to extract the next value of a generator `g`. It simply calls `g()`. If later one wants to pass a C++ iterator where a generator is expected, we can simply reload `next` to take an iterator `it` and return `*it++`. The benefit of having `next` is (a) it makes it easier to spot where the generator is called by simply searching for `next`, and (b) it keeps the simplicity of generator and makes possible the potential use of C++ iterator.

There are two remaining design choices to make:

First one is the order of arguments `Generator &g, T init_val` or `T init_val, Generator &g`. The first one is consistent with the argument order of `std::accumulate`, makes it possible to deduce `T` from `Generator`, and supports a default value. The second one seems more explicit and "logical" (since `init_val` is in front of the series), and is convenient to write if `T` is explicitly specified and `Generator` is deduced.

Second one is error reporting. When `max_terms` have been evaluated without convergence, `series_eval` reports error and returns `nan`, while `series_eval_kahan` returns the last result and a flag indicating non-convergence and lets the caller decide what to do. The latter is more flexible and possibly more readable, especially if the caller has to do something with the result instead of directly returning it. The former is more convenient but the dependence on `special::set_error` doesn't look quite "modular".

What do you think?

steppi commented:
Thanks @fancidev.

> First one is the order of arguments `Generator &g, T init_val` or `T init_val, Generator &g`. The first one is consistent with the argument order of `std::accumulate`, makes it possible to deduce `T` from `Generator`, and supports a default value. The second one seems more explicit and "logical" (since `init_val` is in front of the series), and is convenient to write if `T` is explicitly specified and `Generator` is deduced.

My preference is to have the generator first, the type inferred from the generator, and for `init_val` to default to zero. This is what Boost does for its series evaluators, and the only reason I didn't do that initially is because I wanted to make things simple and easier to review in my first PR. Following the argument order of `std::accumulate` seems most logical to me.

> Second one is error reporting. When `max_terms` have been evaluated without convergence, `series_eval` reports error and returns `nan`, while `series_eval_kahan` returns the last result and a flag indicating non-convergence and lets the caller decide what to do. The latter is more flexible and possibly more readable ("explicit is better than implicit"), especially if the caller has to do something with the result instead of directly returning it. The former is more convenient, but the dependence on `special::set_error` doesn't look very "modular".

I think it's better to return the result and a flag and leave it to the caller to call `set_error` for error handling. You may have noticed that the `series_eval` function needs to take in a string for the function name to pass to `set_error`. We should move away from that.

steppi commented:
Also, @fancidev, can you remove everything from `tools.h` which isn't actually being used in `_iv_ratio`?

fancidev commented:
> Also, @fancidev, can you remove everything from `tools.h` which isn't actually being used in `_iv_ratio`?

Sure, will do.

Also thanks for pointing out the boost series evaluator; I didn’t know there is one! For what it’s worth, boost checks convergence using `<= tol` :-) In fact, it uses `! > tol` , so that the iteration will be cut short if the running result becomes `nan`. Looks like a good trick.

fancidev committed:
Clean up unused code.

fancidev commented:
I have removed the unused code, and made some updates to the argument order, type deduction, default value, etc. Hope it's all right!

I leave the `series_eval` and `series_eval_fixed_length` functions mostly untouched as they are called in several places in `hyp2f1`, and I'd try to limit the impact of this PR to `_iv_ratio`.

fancidev committed:
Rearrange formula to simplify code and improve accuracy.

fancidev commented:
I rearranged the `iv_ratio` continued fraction slightly to simplify code and improve accuracy. Now the peak relative error is $2\epsilon$, and RMSE is around $0.67 \epsilon$, where $\epsilon = 2.2 \times 10^{-16}$. The errors are tested for a few values of $\nu$. For each $\nu$, I test 100,000 points sampled evenly in the _range_ of the function (and then invert them). The result is compared against the "right answer" computed from `mpmath`.

The "series" method with Kahan summation and tolerance $\epsilon/2$ appears to give best accuracy. If using regular summation, the error is roughly doubled. Using $\epsilon$ as tolerance increases the peak error to $2.5 \epsilon$. Using a smaller tolerance has no effect on peak error but slightly increases RMSE.

Using backward evaluation doesn't improve the accuracy.

steppi commented:
Thanks for your patience @fancidev. This is a nice contribution and looks good from a math perspective. I'll finish my review this week and will show you how to make it work with CUDA.

fancidev committed:
Merge branch 'main' of https://github.com/scipy/scipy into iv_ratio

fancidev committed:
Unify definition of inv_ratio(v,x) := iv(v,x)/iv(v-1,x).

fancidev committed:
Add comp argument to return one's complement of iv_ratio.  Adjust c.f. leading term.

fancidev commented:
I made two further small changes (improvements):

(1) Slightly rearranged the evaluation of $R_\nu(x) := I_\nu(x)/I_{\nu-1}(x)$: changed from $R=x/f$ with  $f=x+2\nu+\textrm{c.f.}$, to $R=x/(x+f)$ with $f=2\nu+\textrm{c.f.}$. The main purpose is make it easy to compute $1-R$ (see point 2 below), but the rearranged evaluation also improves the accuracy of $R$ by a small amount. (Peak relative error $2 \epsilon$, RMSE around $0.5 \epsilon$.)

(2) Add a boolean (`int`) argument `comp` (one's complement) to `_iv_ratio`. If non-zero, return $1-I_\nu(x)/I_{\nu-1}(x)$. This is useful for inverting $R_\nu(x) = \rho$ for $\rho$ close to 1 (in which case $R$ has too few useful digits): it helps to reduce the relative error of $R_\nu^{-1}(\rho)$ from a few hundred $\epsilon$ to a few $\epsilon$.

fancidev committed:
Scale multiplier by 2 to avoid subnormal multiplier.

steppi commented on scipy/special/_iv_ratio.h:
```diff
@@ -0,0 +1,96 @@
+    std::uint64_t k_; // current index
+};
+
+inline double iv_ratio(double v, double x, int comp) {
```
Do you anticipate passing arrays for `comp` in the `ufunc`? For other cases in `special`, the ufunc for the complement is a separate function, and I think that's what we should do here to.  I like the way template metaprogramming is used here: https://github.com/scipy/scipy/blob/1457f69ee3acf1093731e724132729b23a0fc3d7/scipy/special/special/wright_bessel.h#L659-L660

to write two very similar functions with slightly different behavior based on a boolean switch.

I guess `iv_ratio_c` would be a reasonable name for the complement for now. We've been discussing a naming overhaul though, so this would likely be changed somewhere down the line.

fancidev replied:
Good idea, a separate function is clearer!

I removed the “comp” argument now to keep things simple for this PR. Another reason is that the relative truncation error of `iv_ratio` is provably small (a few machine epsilon), but the same simple bound doesn’t apply to its complement.

I’ll spend more time to work on the truncation bound of `iv_ratio_c`, and attempt it in a separate PR later, possibly together with `iv_ratio_inv`.

steppi reviewed (changes requested):
This is looking good @fancidev. Even if it's not going in the public API, you'll need to add tests though.

You can compare against `mpmath` (set the `dps` reasonably high). I think it's fine to take around 10-20 reference test cases that you think are reasonably representative. You don't have to compute with `mpmath` in the tests, just use it to create the reference values.

I've also made some minor comments and suggestions.

fancidev committed:
Remove the comp argument of iv_ratio.

A separate function to compute the complement of iv_ratio is
preferred and may be added in the future.

fancidev committed:
Merge branch 'main' of https://github.com/scipy/scipy into iv_ratio

fancidev committed:
Move iv_ratio.h into special/ directory.

fancidev commented on scipy/special/_iv_ratio.h:
Ok will do.

fancidev commented on scipy/special/_iv_ratio.h:
```diff
@@ -0,0 +1,96 @@
+#include "special/error.h"
+#include <cstdint>    // for uint64_t
+#include <cmath>      // for frexp, ldexp, fmax, fma
+#include <utility>    // for pair
```
Noted.

fancidev commented on scipy/special/special/tools.h:
```diff
@@ -73,5 +105,165 @@ namespace detail {
     }
 
+    /* Performs one step of Kahan summation. */
+    template <typename T>
+    SPECFUN_HOST_DEVICE void kahan_step(T *sum, T *comp, T x) {
+        T y = x - *comp;
+        T t = *sum + y;
+        *comp = (t - *sum) - y;
+        *sum = t;
+    }
```
I experimented a few ways to code Kahan summation, such as making a `KahanNumber` class that overloads `+=`, implementing a binary function that looks like `std::plus`, etc. All of these boil down to the computation of what the `kahan_step` function does, so I added that function. I ended up writing `series_eval_kahan` because I find it the most readable.

The benefit of having `kahan_step` as a separate function rather than “inlining” it is (1) for readability, and (2) to make it easier to verify code correctness.

I had a use case for `kahan_step` outside of `series_eval_kahan`. There are different possibilities of choosing `b_0` for the `iv_ratio` continued fraction. I compared three choices: $0$, $2\nu$, and $x+2\nu$. The middle one leads to the lowest RMSE (about 25% lower than the rest), which is intriguing because the first one has a slightly lower theoretical truncation error bound. This is partially resolved by using Kahan summation on the “outside”, though the middle one is still slightly better. So I keep the middle one in the final code.

In summary, I think it’s helpful to keep `kahan_step` as a stand-alone function.

steppi replied:
I see. I personally find it more readable to not have it be a separate function. It's a small enough number of lines that I can mentally chunk it as one thing, and then I find it easier to have everything in one place. Let's just keep it a separate function for now though. Later on, it might make more sense to have `series_eval` take a callable for the summation step which would default to a regular summation, but we could pass in a `kahan_step` or possibly other compensation methods.

fancidev replied:
Ok!

fancidev commented on scipy/special/special/tools.h:
```diff
@@ -8,6 +8,38 @@
+    /* Result type of a "generator", a callable object that produces a value
+     * each time it is called.
+     */
+    template <typename Generator>
+    using generator_result_t = std::decay_t<std::invoke_result_t<Generator>>;
```
Noted thanks.

fancidev commented on scipy/special/_iv_ratio.h:
```diff
@@ -0,0 +1,96 @@
+        // There is not a unique limit as both v and x tends to infinity.
+        special::set_error("iv_ratio", SF_ERROR_DOMAIN, NULL);
+        return std::numeric_limits<double>::signaling_NaN();
+    }
```
I was under the impression to raise a signaling NaN if the function “unexpectedly” fails to return a result, and quiet NaN if the function “expectedly” fails to return a result. In effect, if the user runs a chain of math formula and expects to get a finite result, then they should get an FP exception raised at the first point of invalid operation.

As a benchmark, I find from the “Error handling” section of [sqrt](https://en.cppreference.com/w/c/numeric/math/sqrt) function that it raises FP_INVALID and returns NaN (equivalent to signaling NaN) if the argument is out of domain, but o merely returns NaN if the argument is itself NaN.

steppi replied:
I see, that would make sense in many contexts. What's different here is that the design is such that all error handling goes through [special:set_error](https://github.com/scipy/scipy/blob/main/scipy/special/special/error.h), which let's users control how error handling is done. In SciPy, this delegates to [sf_error](https://github.com/scipy/scipy/blob/main/scipy/special/sf_error.cc), which enables [special.seterr](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.seterr.html), [special.errstate](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.errstate.html#scipy.special.errstate) etc., which provides a Python API to control error handling.

fancidev replied:
Thanks for the explanation. It makes sense now. You also answered my unasked question “how does Python handle FPE”, as it apparently does not catch SIGFPE. Now this becomes clear!

fancidev commented on scipy/special/_iv_ratio.h:
```diff
@@ -0,0 +1,96 @@
+
+    if (terms == 0) { // failed to converge; should not happen
+        special::set_error("iv_ratio", SF_ERROR_NO_RESULT, NULL);
+        return std::numeric_limits<double>::signaling_NaN();
```
Similar to the rationale above, because the iteration “should” converge within so many steps (provable without rounding error, empirical otherwise), I’d like to “warn” the caller should it fail to converge.

Thus, a robust program could set up a signal handler to catch SIGFPE for FP_INVALID and do what it can.

steppi commented on scipy/special/_special_ufuncs.cpp:
```diff
@@ -413,6 +415,12 @@ PyMODINIT_FUNC PyInit__special_ufuncs() {
     );
     PyModule_AddObjectRef(_special_ufuncs, "iv", iv);
 
+    PyObject *iv_ratio = SpecFun_NewUFunc(
+        {static_cast<func_dd_d_t>(special::iv_ratio)},
+        "_iv_ratio", iv_ratio_doc
+    );
```
If you use the new infrastructure, you have to add the `float` loops too, not just `double`. You'll also need to create an overloaded version of the scalar kernel which takes `floats`, casts to double before calling the function, and then casts back to float when returning.

fancidev replied:
Noted thanks.

fancidev commented:
Thanks very much for the thorough review @steppi ! I’ll make some changes accordingly and add some tests.

Btw, I added some implementation notes to the first post of this thread for future reference.

steppi commented:
> Btw, I added some implementation notes to the first post of this thread for future reference.

Thanks. That's very helpful, and when this C++ library we're working on is split out from SciPy the plan is to have detailed implementation notes like that as part of the documentation.

fancidev committed:
Add float overload of iv_ratio().

fancidev committed:
Return quiet NaN as error handling is delegated to set_error().

fancidev committed:
Pass arguments to kahan_step() by reference.

fancidev committed:
Return signed zero if x is signed zero.

fancidev committed:
Add unit tests for iv_ratio.

fancidev committed:
Merge branch 'main' of https://github.com/scipy/scipy into iv_ratio

fancidev committed:
Ignore spurious mypy error in unit test file.

fancidev commented:
@steppi I added unit tests, as well as addressed all the other points in the code review. Glad to declare the code now “bullet-proof”!

Would you please review again when you have time?

steppi reviewed (approved):
Very nice work @fancidev. I think this is ready to merge. You've done some things which won't work in CuPy, but I'll take care of that in a follow-up, since you mentioned you don't have access to a machine with an NVIDIA GPU.

fancidev commented:
Thank you very much @steppi for all the review and guide!

I’ll work on the inverse of `iv_ratio` next. After that, I’ll update `vonmises` and `vonmises_fisher` to make use of these.
